app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -5,106 +5,197 @@ import android.graphics.Bitmap
     }
 
-    fun getBitmap(): Bitmap? = maskBitmap
+    class CustomViewBaseState(
+        pathsAndPaints: List<Pair<PointToPointPath, Paint>>,
+        val undoPendingPaths: Int
+    ) : Parcelable {
 
-    companion object {
-        private const val MATRIX_SIZE = 9
-        private const val MASK_PAINT_ALPHA = 200
+        companion object CREATOR : Parcelable.Creator<CustomViewBaseState> {
+            override fun createFromParcel(parcel: Parcel): CustomViewBaseState {
+                return CustomViewBaseState(parcel)
+            }
+
+            override fun newArray(size: Int): Array<CustomViewBaseState?> {
+                return arrayOfNulls(size)
+            }
+        }
+
+        val basePathsAndPaintChangesData: MutableList<Pair<PathBaseData, PaintChangeBaseData?>> = LinkedList()
+
+        init {
+            if (pathsAndPaints.isNotEmpty()) {
+                val (firstPath, firstPaint) = pathsAndPaints[0]
+                val firstPathBaseData = PathBaseData(firstPath.getPoints())
+                val firstPaintChangeBaseData =
+                    PaintChangeBaseData(firstPaint.color, firstPaint.strokeWidth)
+                basePathsAndPaintChangesData.add(firstPathBaseData to firstPaintChangeBaseData)
+                var latestPaint: Paint = firstPaint
+                for ((path, paint) in pathsAndPaints.subList(1, pathsAndPaints.size)) {
+                    val pathPoints = path.getPoints()
+                    val pathBaseData = PathBaseData(pathPoints)
+                    val paintBaseChangeData =
+                        if (paint != latestPaint)
+                            PaintChangeBaseData(paint.color, paint.strokeWidth)
+                        else
+                            null
+                    basePathsAndPaintChangesData.add(pathBaseData to paintBaseChangeData)
+                    latestPaint = paint
+                }
+            }
+        }
+
+        constructor(parcel: Parcel) : this(LinkedList<Pair<PointToPointPath, Paint>>().apply {
+            parcel.readList(this as List<*>, Pair::class.java.classLoader)
+        }, parcel.readInt())
+
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            parcel.writeList(basePathsAndPaintChangesData as List<*>)
+            parcel.writeInt(undoPendingPaths)
+        }
+
+        override fun describeContents() = 0
+
+        fun composePathsAndPaints(): List<Pair<PointToPointPath, Paint>> {
+            val result = LinkedList<Pair<PointToPointPath, Paint>>()
+            lateinit var latestPaint: Paint
+            for ((basePath, paintChange) in basePathsAndPaintChangesData) {
+                val path = PointToPointPath(basePath.points)
+                latestPaint =
+                    if (paintChange == null)
+                        latestPaint
+                    else
+                        newDefaultPaintBrush(paintChange.color).apply {
+                            strokeWidth = paintChange.strokeWidth
```

Comment:
I noticed that you always set the `strokeWidth` in an `apply` block after calling `newDefaultPaintBrush()`, why not add it as a parameter of the said method?